packages/protoplugin/src/ecmascript/schema.ts

```diff
@@ -216,9 +216,19 @@ function getFilesToGenerate(
       }
     }
   }
+  // Our goal is to provide options with source retention to plugin authors.
+  // CodeGeneratorRequest.proto_file elides options with source retention for
+  // files to generate. For these files, we take the file from source_file_descriptors,
+  // which does include options with source retention.
+  const allProtoWithSourceOptions = request.protoFile.map((protoFile) => {
+    const sourceFile = request.sourceFileDescriptors.find(
+      (s) => s.name == protoFile.name,
+    );
+    return sourceFile ?? protoFile;
+  });
```

Comment:
This is backwards compatible. Old compilers will not populate `source_file_descriptors`, and we keep using `proto_file`.